stats/opentelemetry/internal/grpc_trace_bin_propagator.go

```diff
@@ -0,0 +1,139 @@
+// text format (slow path). If span context is not valid or emptu, no data is
+// injected.
+func (p GRPCTraceBinPropagator) Inject(ctx context.Context, carrier otelpropagation.TextMapCarrier) {
+	span := oteltrace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return
+	}
+
+	bd := binary(span.SpanContext())
+	if bd == nil {
+		return
+	}
+
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		cc.SetBinary(bd)
+		return
+	}
+	carrier.Set(itracing.GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(bd))
+}
+
+// Extract reads OpenTelemetry trace context information from the carrier into a
+// Context.
+//
+// If the carrier is a CustomCarrier, trace data is read directly in a binary
+// format from the `grpc-trace-bin` header (fast path). Otherwise, the trace
+// data is base64 decoded from the same header in text format (slow path).
+//
+// If a valid trace context is found, this function returns a new context
+// derived from the input `ctx` containing the extracted span context. The
+// extracted span context is marked as "remote", indicating that the trace
+// originated from a different process or service. If trace context is invalid
+// or not present, input `ctx` is returned as is.
+func (p GRPCTraceBinPropagator) Extract(ctx context.Context, carrier otelpropagation.TextMapCarrier) context.Context {
+	var bd []byte
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		bd = cc.GetBinary()
+	} else {
+		bd, _ = base64.StdEncoding.DecodeString(carrier.Get(itracing.GRPCTraceBinHeaderKey))
+	}
+	if bd == nil {
+		return ctx
+	}
+
+	sc, ok := fromBinary([]byte(bd))
+	if !ok {
+		return ctx
+	}
+	return oteltrace.ContextWithRemoteSpanContext(ctx, sc)
+}
+
+// Fields returns the keys whose values are set with Inject.
+//
+// GRPCTraceBinPropagator always returns a slice containing only
+// `grpc-trace-bin` key because it only sets the `grpc-trace-bin` header for
+// propagating trace context.
+func (p GRPCTraceBinPropagator) Fields() []string {
+	return []string{itracing.GRPCTraceBinHeaderKey}
+}
+
+// Binary returns the binary format representation of a SpanContext.
+//
+// If sc is the zero value, returns nil.
+func binary(sc oteltrace.SpanContext) []byte {
+	if sc.Equal(oteltrace.SpanContext{}) {
+		return nil
+	}
+	var b [29]byte
+	traceID := oteltrace.TraceID(sc.TraceID())
+	copy(b[2:18], traceID[:])
+	b[18] = 1
+	spanID := oteltrace.SpanID(sc.SpanID())
+	copy(b[19:27], spanID[:])
+	b[27] = 2
+	b[28] = byte(oteltrace.TraceFlags(sc.TraceFlags()))
+	return b[:]
+}
+
+// FromBinary returns the SpanContext represented by b with Remote set to true.
+//
+// It returns with zero value SpanContext and false, if any of the
+// below condition is not satisfied:
+// - Valid header: len(b) = 29
+// - Valid version: b[0] = 0
+// - Valid traceID prefixed with 0: b[1] = 0
+// - Valid spanID prefixed with 1: b[18] = 1
+// - Valid traceFlags prefixed with 2: b[27] = 2
+func fromBinary(b []byte) (oteltrace.SpanContext, bool) {
```

Comment:
Optional: Should this return an error instead? Or a nil span context and return a *?

But this is probably fine I think.

Reply:
yeah i am ok with it. Also, we have test for each case